src/hotspot/share/cds/archiveHeapWriter.cpp

```diff
@@ -205,8 +205,13 @@ void ArchiveHeapWriter::copy_roots_to_buffer(GrowableArrayCHeap<oop, mtClassShar
   memset(mem, 0, byte_size);
   {
     // This is copied from MemAllocator::finish
-    oopDesc::set_mark(mem, markWord::prototype());
-    oopDesc::release_set_klass(mem, k);
+    if (UseCompactObjectHeaders) {
+      narrowKlass nk = ArchiveBuilder::current()->get_requested_narrow_klass(k);
+      oopDesc::release_set_mark(mem, markWord::prototype().set_narrow_klass(nk));
+    } else {
+      oopDesc::set_mark(mem, markWord::prototype());
+      oopDesc::release_set_klass(mem, k);
+    }
```

Comment:
The `UseCompactObjectHeaders` path calls `get_requested_narrow_klass`, while the `else` part directly uses `k`. Is one of these paths incorrect?

Reply:
This seems odd. The original code sets `Universe::objectArrayKlass()` into the object header. This is the value of this class in the current JVM lifetime.

Later, `ArchiveHeapWriter::update_header_for_requested_obj()` would change the object's klass to the "requested" address. I.e., where this class will be loaded in a future JVM lifetime when the CDS archive is loaded into memory.

It seems the same logic should be used in the `UseCompactObjectHeaders==true` case.

BTW (unrelated to this PR) the comment a few lines up is outdated and wrong:

```
Klass* k = Universe::objectArrayKlass(); // already relocated to point to archived klass
```

`k` is the value of the *actual* location of this class in the current JVM lifetime. Please ignore this comment when trying to understand this function.